### ISSUE TITLE
A question about loss backward

### ISSUE THREAD

**Issue by QWTforGithub** [jacobgil/pytorch-grad-cam#135]
Thank you for sharing the well organized code. I learned a lot from the code. Now I would like to ask you a question about loss.backward. For the classification problem, the final output of the model is a classification vector, such as 1 x M vector, your code is also targeted at the classification task, for a certain category, one of the categories corresponding to the classification score can be propagated backward. My question is, if the output of the model is not a 1 x M vector, but a matrix like N x M, and I want to take one of N and propagate backwards, how do I do that? 

   Your code: 1 x M -> value
   I want: N x M -> a vector

Looking forward to your reply.

**Comment by jacobgil** [comment]
Hi,
I think the first thing to clarify here is what would be the target of the visualization.

For the usual classification case, the visualizations answer the question "What is the attribution for category C?".

For the vector case, what is it that the visualization should answer?
For example:
- Maybe you want to aggregate it by taking the sum? Maybe you want to take the similarity of this vector from some other vector?
- Maybe you want to create N different visualizations, one for every element in the vector?

**Comment by QWTforGithub** [comment]
> Hi,
> I think the first thing to clarify here is what would be the target of the visualization.
> 
> For the usual classification case, the visualizations answer the question "What is the attribution for category C?".
> 
> For the vector case, what is it that the visualization should answer?
> For example:
> 
> * Maybe you want to aggregate it by taking the sum? Maybe you want to take the similarity of this vector from some other vector?
> * Maybe you want to create N different visualizations, one for every element in the vector?
Thank you for your reply. My task is not a categorization task, but the model I want to visualize is a descriptor that generates some descriptor in the network. The descriptor network outputs a matrix of N x M descriptors, where N represents the length of the descriptor and M represents the number of descriptors. I want to take one of the descriptors as the target and propagate backwards so that I can get the points that affect the target descriptor and visualize them. But since this is not a classification task, and the descriptor is a vector, not a value, how do I backward propagate a vector, or find the path that this vector travels backward? 

Thank you for your reply.

**Comment by QWTforGithub** [comment]
> Hi,
> I think the first thing to clarify here is what would be the target of the visualization.
> 
> For the usual classification case, the visualizations answer the question "What is the attribution for category C?".
> 
> For the vector case, what is it that the visualization should answer?
> For example:
> 
> * Maybe you want to aggregate it by taking the sum? Maybe you want to take the similarity of this vector from some other vector?
> * Maybe you want to create N different visualizations, one for every element in the vector?

Thank you for your reply. My task is not a categorization task, but the model I want to visualize is a descriptor that generates some descriptor in the network. The descriptor network outputs a matrix of N x M descriptors, where N represents the length of the descriptor and M represents the number of descriptors. I want to take one of the descriptors as the target and propagate backwards so that I can get the points that affect the target descriptor and visualize them. But since this is not a classification task, and the descriptor is a vector, not a value, how do I backward propagate a vector, or find the path that this vector travels backward?

Thank you for your reply.

**Comment by QWTforGithub** [comment]
> Hi,
> I think the first thing to clarify here is what would be the target of the visualization.
> 
> For the usual classification case, the visualizations answer the question "What is the attribution for category C?".
> 
> For the vector case, what is it that the visualization should answer?
> For example:
> 
> * Maybe you want to aggregate it by taking the sum? Maybe you want to take the similarity of this vector from some other vector?
> * Maybe you want to create N different visualizations, one for every element in the vector?

Thank you for your reply. My task is not a categorization task, but the model I want to visualize is a descriptor that generates some descriptor in the network. The descriptor network outputs a matrix of N x M descriptors, where N represents the length of the descriptor and M represents the number of descriptors. I want to take one of the descriptors as the target and propagate backwards so that I can get the points that affect the target descriptor and visualize them. But since this is not a classification task, and the descriptor is a vector, not a value, how do I backward propagate a vector, or find the path that this vector travels backward?

**Comment by jacobgil** [comment]
I think you can backpropagate the vector, and the gradient result will a vector - the gradient vs respect to every element in the garget vector.

But to do the visualization, you will still need to decide what is the target you want to visualize.
Otherwise if you don't aggregate the descriptor based on some target function, you will get "many" visualizations - one for every different output element in the target descriptor. Is that what you want?

Every element in your descriptor might have a very negative on positive number. When you making this kind of a visualization, you're trying to answer questions like "what makes this element more positive?" -> But you need to decide what is the exact question you want to answer.

If you have some reference category descriptor, e.g the desctiptor of a cat, maybe you can create the descriptor by taking the  cosine similarity with the "cat" descriptor.

**Comment by jacobgil** [comment]
Btw - is this based on some standard architecture, maybe deep learning based keypoint detection?
It could be nice to add an example like this to the repository so I can help if you share more details.

**Comment by QWTforGithub** [comment]
> I think you can backpropagate the vector, and the gradient result will a vector - the gradient vs respect to every element in the garget vector.
> 
> But to do the visualization, you will still need to decide what is the target you want to visualize.
> Otherwise if you don't aggregate the descriptor based on some target function, you will get "many" visualizations - one for every different output element in the target descriptor. Is that what you want?
> 
> Every element in your descriptor might have a very negative on positive number. When you making this kind of a visualization, you're trying to answer questions like "what makes this element more positive?" -> But you need to decide what is the exact question you want to answer.
> 
> If you have some reference category descriptor, e.g the desctiptor of a cat, maybe you can create the descriptor by taking the cosine similarity with the "cat" descriptor.

Thank you for your unselfish advice. I received it. As you said, the positive value of an element of the descriptor may cause a negative effect. I will consider this issue carefully. Thanks again!

**Comment by jacobgil** [comment]
Great.
Closing this for now, please ask any other questions that come up.
